### DISCUSSION TITLE
Problem with Setting up Ziggy Laravel9/Vue3/inertia/vite

### DISCUSSION THREAD

**Discussion by Lx45** [tighten/ziggy#565] [Q&A] [answered]
I'm really desperate for help, I cant figure it out on my own.

I try to setup Ziggy.js in my app. -> https://github.com/tighten/ziggy

First I install ziggy via

`composer require tightenco/ziggy`

After that I Add the @routes Blade directive to my main layout (Don't know if this is necessary) 

app.blade.php
```
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<link rel="preconnect" href="https://fonts.googleapis.com">
		<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
		<link href="https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,500;0,700;1,500&display=swap" rel="stylesheet">
		@routes
		<title>Laravel with Inertia</title>
		@vite
	</head>
	<body class="antialiased bg-gray-900">
		@inertia
	</body>
</html>
```

After that I generate the JS file with `php artisan ziggy:generate`

Next I create an alias.

vite.config.js
```
import { defineConfig } from 'vite'
import tailwindcss from 'tailwindcss'
import autoprefixer from 'autoprefixer'
import laravel from 'vite-plugin-laravel'
import vue from '@vitejs/plugin-vue'
import inertia from './resources/scripts/vite/inertia-layout'
import path from 'path'

export default defineConfig({
	plugins: [
		inertia(),
		vue(),
		laravel({
			postcss: [
				tailwindcss(),
				autoprefixer(),
			],
		}),
	],
	resolve: {
		alias: {
			ziggy: path.resolve('vendor/tightenco/ziggy/dist'),
		},
	},
})
```

Then I add the Import to my main.ts file

```
import { createApp, h } from 'vue'
import route from 'ziggy';
import { Ziggy } from './ziggy';
import { createInertiaApp } from '@inertiajs/inertia-vue3'
import { importPageComponent } from '@/scripts/vite/import-page-component'

createInertiaApp({
	resolve: (name) => importPageComponent(name, import.meta.glob('../views/pages/**/*.vue')),
	setup({ el, app, props, plugin }) {
		createApp({ render: () => h(app, props) })
			.use(plugin,Ziggy)
			.mount(el)
	},
})
```

And I get the following error 

```
[plugin:vite:import-analysis] Failed to resolve import "./ziggy" from "resources/scripts/main.ts". Does the file exist?
/Users/lx/Code/focus/resources/scripts/main.ts:3:0
1  |  import { createApp, h } from "vue";
2  |  import { Ziggy } from "./ziggy";
   |                         ^
3  |  import { createInertiaApp } from "@inertiajs/inertia-vue3";
4  |  import { importPageComponent } from "@/scripts/vite/import-page-component";
```

**Comment by bakerkretzmar** [comment, marked as answer]
Does the `ziggy.js` file exist right beside your `main.ts`? If it does, does importing it as `'./ziggy.js'` (including file extension) work?

Also, your `.use()` for the Vue plugin should have Ziggy's Vue plugin as its first argument, not the route function.

```diff
- import route from 'ziggy';
+ import { ZiggyVue } from 'ziggy'; // if this doesn't work, try importing it from `ziggy/dist/vue`

- .use(plugin,Ziggy)
+ .use(ZiggyVue, Ziggy)
```

Take a look at [the docs for using the Vue plugin](https://github.com/tighten/ziggy#vue).

**Reply by Lx45** [reply]
Thank you for your help.

I moved the Ziggy.js file next to the main.ts file. And updated the path like you described. 

Path to the maint.ts file and ziggy.js file -> 
resources/scripts/main.ts
resources/scripts/ziggy.js

If  I do it like this
```
import { createApp, h } from 'vue'
import { ZiggyVue } from 'ziggy';
import { Ziggy } from './ziggy.js';
import { createInertiaApp } from '@inertiajs/inertia-vue3'
import { importPageComponent } from '@/scripts/vite/import-page-component'

createInertiaApp({
	resolve: (name) => importPageComponent(name, import.meta.glob('../views/pages/**/*.vue')),
	setup({ el, app, props, plugin }) {
		createApp({ render: () => h(app, props) })
			.use(plugin,ZiggyVue, Ziggy)
			.mount(el)
	},
})
```

I get the following error
```
Uncaught SyntaxError: The requested module '/vendor/tightenco/ziggy/dist/index.js' does not provide an export named 'ZiggyVue' (at main.ts:2:1)
```

If i add the path to the dist folder like this 
```
import { createApp, h } from 'vue'
import { ZiggyVue } from 'ziggy/dist/vue';
import { Ziggy } from './ziggy.js';
import { createInertiaApp } from '@inertiajs/inertia-vue3'
import { importPageComponent } from '@/scripts/vite/import-page-component'

createInertiaApp({
	resolve: (name) => importPageComponent(name, import.meta.glob('../views/pages/**/*.vue')),
	setup({ el, app, props, plugin }) {
		createApp({ render: () => h(app, props) })
			.use(plugin,ZiggyVue, Ziggy)
			.mount(el)
	},
})
```

I get this error 
```
GET http://localhost:3000/vendor/tightenco/ziggy/dist/dist/vue net::ERR_ABORTED 404 (Not Found)
```

My application runs on http://127.0.0.1:8000

fyi this is how my ziggy.js file looks like
```
const Ziggy = {"url":"http:\/\/focus.test","port":null,"defaults":{},"routes":{"ignition.healthCheck":{"uri":"_ignition\/health-check","methods":["GET","HEAD"]},"ignition.executeSolution":{"uri":"_ignition\/execute-solution","methods":["POST"]},"ignition.updateConfig":{"uri":"_ignition\/update-config","methods":["POST"]},"dashboard":{"uri":"dashboard","methods":["GET","HEAD"]},"users.create":{"uri":"users\/create","methods":["GET","HEAD"]},"users.store":{"uri":"users","methods":["POST"]},"session.create":{"uri":"session\/create","methods":["GET","HEAD"]}}};

if (typeof window !== 'undefined' && typeof window.Ziggy !== 'undefined') {
    Object.assign(Ziggy.routes, window.Ziggy.routes);
}

export { Ziggy };
```

**Reply by Lx45** [reply]
If I import it like this 
```
import { createApp, h } from 'vue'
import { ZiggyVue } from '/vendor/tightenco/ziggy/dist/vue';
import { Ziggy } from './ziggy.js';
import { createInertiaApp } from '@inertiajs/inertia-vue3'
import { importPageComponent } from '@/scripts/vite/import-page-component'

createInertiaApp({
	resolve: (name) => importPageComponent(name, import.meta.glob('../views/pages/**/*.vue')),
	setup({ el, app, props, plugin }) {
		createApp({ render: () => h(app, props) })
			.use(plugin,ZiggyVue, Ziggy)
			.mount(el)
	},
})
```

I get the error
```
Uncaught SyntaxError: The requested module '/vendor/tightenco/ziggy/dist/vue.js' does not provide an export named 'ZiggyVue' (at main.ts:2:1)
```

**Reply by bakerkretzmar** [reply]
Okay, you may need to mess around with the import path for `ZiggyVue` some more, I'm not exactly sure what it should be with Vite. Try just `'ziggy'` again but change the alias in your config to `vendor/tightenco/ziggy/dist/vue`, or `vendor/tightenco/ziggy/dist/vue.js`, or `vendor/tightenco/ziggy/dist/vue.es.js`. If that doesn't work, remove the alias completely and try importing from the complete relative path from your `main.ts` to the `dist/vue.js` or `dist/vue.es.js` file.

Your `.use()` is also still incorrect, it should just be exactly `.use(ZiggyVue, Ziggy)`.

**Reply by Lx45** [reply]
Yes! It works now 🚀 

Thank you 🙏 

I imported it as just `ziggy` and changed the alias to `vendor/tightenco/ziggy/dist/vue.es.js`

**Reply by bakerkretzmar** [reply]
Great! Let me know if you run into anything else 💪🏻

**Reply by khalidmaquilang** [reply]
do i need to run `php artisan ziggy:generate` whenever there is a new route?

**Reply by bakerkretzmar** [reply]
@khalidmaquilang yes, because that file contains a list of all your routes.

**Reply by khalidmaquilang** [reply]
why is it in laravel jetstream, we don't need to run that command

**Reply by bakerkretzmar** [reply]
@khalidmaquilang there are three different ways to load your Ziggy config—using the `@routes` Blade directive to output it in a view, running `php artisan ziggy:generate` to store it in a file, or retrieving it manually at runtime.

If you use `php artisan ziggy:generate`, you have to re-run that command to update the file whenever your routes change, and you have to import the file and pass it into the `route()` function or the Vue plugin.

Depending which preset/stack you use, Jetstream either uses the `@routes` Blade directive or shares the routes with Inertia at runtime in a middleware. See https://github.com/search?q=repo%3Alaravel%2Fjetstream+ziggy&type=code.

The original example in this discussion uses both `@routes` and `php artisan ziggy:generate`, which isn't necessary.

**Reply by khalidmaquilang** [reply]
@bakerkretzmar i followed the jetstream way but im receiving Ziggy is not defined

**Reply by bakerkretzmar** [reply]
I'm happy to try to help but you need to share as much information as possible. Please reply or open a new issue/discussion with complete code examples of what you're doing, your Ziggy config and app setup, and exactly the error you're getting.

**Reply by khalidmaquilang** [reply]
@bakerkretzmar soon.. i just used their command and install jetstream automatically.. ill just post another issue if i will encounter that problem in my future projects.. thank you!

**Reply by buzdyk** [reply]
solution that worked for me (Laravel 10, Vue 3):

Composer dependency: `"tightenco/ziggy": "^1.0"` (not 2!!)

app.js :
```
import { ZiggyVue } from '../../vendor/tightenco/ziggy/dist/vue.m'

Vue.use(ZiggyVue)
```

**Reply by sbaum2s** [reply]
> solution that worked for me (Laravel 10, Vue 3):
> 
> Composer dependency: `"tightenco/ziggy": "^1.0"` (not 2!!)
> 
> app.js :
> 
> ```
> import { ZiggyVue } from '../../vendor/tightenco/ziggy/dist/vue.m'
> 
> Vue.use(ZiggyVue)
> ```

In version 1 there is no file like "vue" in dist folder.

**Comment by AMINEAMIRI** [comment]
hii,
i have the same probleme in deploy with cyclicapp thats the error ; Error: Could not load /tmp/myproject/vendor/tightenco/ziggy/dist/vue.es (imported by resources/js/app.js): ENOENT: no such file or directory, open '/tmp/myproject/vendor/tightenco/ziggy/dist/vue.es',

`app.js
import "./bootstrap";
import "../css/app.css";

import { createApp, h } from "vue";
import { createInertiaApp } from "@inertiajs/inertia-vue3";
import { InertiaProgress } from "@inertiajs/progress";
import { resolvePageComponent } from "laravel-vite-plugin/inertia-helpers";
import { ZiggyVue } from "ziggy";

const appName =
    window.document.getElementsByTagName("title")[0]?.innerText || "Laravel";

createInertiaApp({
    title: (title) => `${title} - ${appName}`,
    resolve: (name) =>
        resolvePageComponent(
            `./Pages/${name}.vue`,
            import.meta.glob("./Pages/**/*.vue")
        ),
    setup({ el, app, props, plugin }) {
        return createApp({ render: () => h(app, props) })
            .use(plugin)
            .use(ZiggyVue, Ziggy)
            .mount(el);
    },
});`

`package.json:


    "scripts": {
        "dev": "vite",
        "build": "vite build",
        "start": "node vite.config.js"
    },`

`vite.config.js:
import { defineConfig } from "vite";
import laravel from "laravel-vite-plugin";
import vue from "@vitejs/plugin-vue";
import path from "path";

export default defineConfig({
    plugins: [
        laravel({
            input: "resources/js/app.js",
            refresh: true,
        }),
        vue({
            template: {
                transformAssetUrls: {
                    base: null,
                    includeAbsolute: false,
                },
            },
        }),
    ],
    resolve: {
        alias: {
            ziggy: path.resolve("vendor/tightenco/ziggy/dist/vue.es"),
        },
    },
});
`

Any Help please

**Reply by bakerkretzmar** [reply]
There's no file `vue.es`, try changing the alias path so it ends with either the 'module' path to just the base filename, `vue`, or the full path to a specific file, e.g. `vue.es.js`.

**Reply by AMINEAMIRI** [reply]
still have the same probleme , any other solutions ? thank you alot

[vite:load-fallback] Could not load /tmp/myproject/vendor/tightenco/ziggy/dist/vue.es.js (imported by resources/js/app.js): ENOENT: no such file or directory, open '/tmp/myproject/vendor/tightenco/ziggy/dist/vue.es.js'
error during build:
Error: Could not load /tmp/myproject/vendor/tightenco/ziggy/dist/vue.es.js (imported by resources/js/app.js): ENOENT: no such file or directory, open '/tmp/myproject/vendor/tightenco/ziggy/dist/vue.es.js'

**Reply by bakerkretzmar** [reply]
Does the `/tmp/myproject/vendor/tightenco/ziggy/dist/vue.es.js` file exist?

**Reply by AMINEAMIRI** [reply]
yes the file exist , in local it works fine , but whene i try to deploy the project with cyclicapp it gives me this error

**Reply by bakerkretzmar** [reply]
I don't know what cyclicapp is, are you sure the file exists during the deployment and in production? I need more information to try to help figure out what's going on, but it sounds like a configuration/deployment issue if it works locally.

**Reply by AMINEAMIRI** [reply]
cyclicapp is a pass like heroku, thank you very much for your help I will try to find another free fullstack PASS

**Comment by craigrileyuk** [comment]
[AdminUI: Inertia Routes](https://github.com/AdminUI/inertia-routes)

Company I work for made this to streamline our use of Inertia with named routes, MIT and thought it might be useful to some of you all.

Gets you up and running with named routes in Inertia/Vue3 with around 4 lines of code, no @routes directive, exported JS files or API routes setup needed.
